src/prompt.ts

```diff
@@ -0,0 +1,76 @@
+import * as inquirer from "inquirer";
+import * as _ from "lodash";
+
+import * as FirebaseError from "./error";
+
+export interface Choice {
+  checked: boolean;
+  label: string;
+  name: string;
+}
+export type Question = inquirer.Question;
+
+/**
+ * prompt is used to prompt the user for values. Specifically, any `name` of a
+ * provided question will be checked against the `options` object. If `name`
+ * exists as a key in `options`, it will *not* be prompted for. If `options`
+ * contatins `nonInteractive = true`, then any `question.name` that does not
+ * have a value in `options` will cause an error to be returned. Once the values
+ * are queried, the values for them are put onto the `options` object, and the
+ * answers are returned.
+ * @param options The options object passed through by Command.
+ * @param questions `Question`s to ask the user.
+ * @return The answers, keyed by the `name` of the `Question`.
+ */
+export async function prompt(options: { [key: string]: any }, questions: Question[]): Promise<any> {
+  const prompts = [];
+  for (const question of questions) {
+    if (question.name && !options[question.name]) {
+      prompts.push(question);
+    }
+  }
+
+  if (prompts.length && options.nonInteractive) {
+    throw new FirebaseError(
+      "Missing required options (" +
+        _.uniq(_.map(prompts, "name")).join(", ") +
+        ") while running in non-interactive mode",
+      {
+        children: prompts,
+        exit: 1,
+      }
+    );
+  }
+
+  const answers = await inquirer.prompt(prompts);
+  _.forEach(answers, (v, k) => {
+    options[k] = v;
```

Comment:
I'm having a hard time understanding this - answers is a key value hash, with key being `question.name` and value being the input that was collected for that question. The odd thing to me is we're passing a callback that takes value as the first param and key as the second param? Then reversing their order when we set options?

Reply:
This is correct. See the documentation for `_.forEach` - the callback is `(value, key)`. This makes it so that the "keys" are being set with the  "values" (answers). Added a comment for clarity.